### DISCUSSION TITLE
Open e2e Testing Initiative

### DISCUSSION THREAD

**Discussion by unlikelyzero** [nasa/openmct#4303] [Announcements]
Follow this Discussion to learn about the Open e2e Testing Initiative with Open MCT!

# Goals
1. Increase our e2e testing coverage 
2. Provide an opportunity for the testing community to directly contribute to NASA Open Source
3. Learn from folks who demonstrate some new approaches to old testing challenges with Playwright.
4. Ensure that Open MCT testing is accessible to everyone

# How to contribute
We will be open sourcing and outlining new e2e tests that will follow our [existing contributing guidelines](https://github.com/nasa/openmct/discussions/3821).

**Please make a comment** in this Github Discussion if you are interested in participating and include if you've already submitted your CLA!

### Getting started with Playwright and Open MCT
[Getting started with Playwright and Open MCT](https://github.com/nasa/openmct/tree/master/e2e#getting-started)

Don't worry if you haven't submitted your CLA, you can get started without it. It'll only prevent the final merge. **You should still submit this ASAP as it can take up to 3 weeks.**

# What to contribute
After finishing *How to contribute* above, there are two ways of contributing tests:

## Help Wanted tickets in Improve Test Coverage
1. Complete the 'How to Contribute' section above
2. Check the [list of help_wanted issues on the Improve Test Coverage Project](https://github.com/nasa/openmct/issues?q=is%3Aissue+label%3Aneeds%3Ae2e+project%3Anasa%2Fopenmct%2F44+label%3Ahelp_wanted+). 
3. If you find one that you want to work on, please assign it to yourself and update your Discussion comment thread that you've taken it on! 
5. If you have any additional questions about these, please ask within a new thread on this Github Discussion.

## Test TODOs within the codebase (aka test.fixme())
1. Complete the 'How to Contribute' section above
2. Check our codebase for tests labeled with the `test.fixme()`. 
3. If you find one, update your Github Discussion and we'll make a new Issue to track the work. 
The `test.fixme()` lines are found within the `e2e/tests/**.e2e.spec.js` files and look like this:

![Screen Shot 2022-03-09 at 6 38 18 AM](https://user-images.githubusercontent.com/12446869/157463394-7ca19b33-68d4-4f9e-bf87-3b51b57b32c4.png)

## Something else?
If you have an idea about what you'd want to contribute, but don't see it list, please make a thread on this discussion!

## Current Contributors

@gholin - CLA Approved - Tests Implemented
@afahey87 - CLA Approved - Tests Implemented
@dimitriharding - CLA Approved - Tests Implemented
@loftwah - CLA Approved - Tests Implemented
@MarkCodering - CLA ?

**Comment by afahey87** [comment]
Hi John, I am interested in participating, I watched your keynote address at StarEast. II'll get started on the [Contributor License Agreement (CLA).

**Reply by afahey87** [reply]
I have emailed the CLA.

**Reply by unlikelyzero** [reply]
@afahey87 are you still at STAREAST?

**Reply by afahey87** [reply]
No I am at home in Maine I was not able to go this year. I watched your Keynote virtually,  It was awesome! I hope to be at STAREAST next year. I am going to talk to my manager about STARWEST maybe that will be in my cards for later this year.

**Reply by unlikelyzero** [reply]
> No I am at home in Maine I was not able to go this year. I watched your Keynote virtually,  It was awesome! I hope to be at STAREAST next year. I am going to talk to my manager about STARWEST maybe that will be in my cards for later this year. 

Great! I'll reach out with some quick and easy tests to get started!

**Reply by afahey87** [reply]
That sounds great! I cant wait to start. :)

**Reply by unlikelyzero** [reply]
@afahey87 I found the perfect test suite to add and get started https://github.com/nasa/openmct/issues/4914#issuecomment-1116678441
Please let me know if you have any difficulties getting started. I'm going to write a quick start guide for e2e testing right now

**Reply by afahey87** [reply]
Thanks Ill take a look at #4914. Let me know when you have the quick start guide up, I would love to have a reference point. Thanks

**Reply by unlikelyzero** [reply]
@afahey87 great work! Now that you're familiar with our codebase, do you see any other fixme() that you want to take a stab at?

**Reply by afahey87** [reply]
@unlikelyzero maybe this one? https://github.com/nasa/openmct/blob/492ff2fa64274b1cf711e7eea0ec0aef9b258df8/e2e/tests/persistence/persistability.e2e.spec.js

**Reply by unlikelyzero** [reply]
@afahey87 actually... I think that one may have been addressed as a part of https://github.com/nasa/openmct/pull/5128

Hmmmm

**Reply by afahey87** [reply]
No problem. Ill take a look again later tonight.

**Reply by afahey87** [reply]
@unlikelyzero How about this one? https://github.com/nasa/openmct/blob/ab4d0dd37f2269de84b6f8f4b6be0f0874018f57/e2e/tests/plugins/imagery/exampleImagery.e2e.spec.js

**Reply by unlikelyzero** [reply]
@afahey87 that would be incredible :)

**Reply by afahey87** [reply]
Challenge accepted! 🤓

**Comment by dimitriharding** [comment]
Hi John, I am very interested in contributing to this project that you mentioned during your keynote at StarEast.

If you could let me know where I can start, that would be great!

**Reply by unlikelyzero** [reply]
@dimitriharding I found the perfect test to get started with!

https://github.com/nasa/openmct/issues/4667#issuecomment-1118032595

**Reply by dimitriharding** [reply]
Okay, nice! I'll start there.

**Reply by dimitriharding** [reply]
@unlikelyzero I believe I have completed the work for #4667, and it's currently on a branch in my forked project. Also, I sent off the IND CLA today.

Can I create the PR and get feedback or should I wait until my CLA gets approved?

**Reply by unlikelyzero** [reply]
@dimitriharding you can make a PR for work and we can complete the code review. We can't merge the work until HQ accepts it. They're now processing them in under a week.

**Reply by dimitriharding** [reply]
@unlikelyzero Can I have a go at [exportASJson](https://github.com/nasa/openmct/blob/master/e2e/tests/plugins/ExportAsJSON/exportAsJson.e2e.spec.js)

Is there a particular issue I need to reference or I can make the PR without an issue reference once done?

**Reply by unlikelyzero** [reply]
@dimitriharding just making a comment in this thread is enough! I want to avoid tickets for these as our issue backlog is big enough as it is. 

After some more thought, maybe we could just remove the importAsJSON and ExportAsJSON folders and make a single testsuite in the root directory which:

1. Creates an Object with some properties
2. Exports as JSON
3. (Optional) Verifies the File in some way. Maybe do deep inspection of the JSON file?
4. Imports that same JSON
5. Verify that the Domain object appears the same.

What do you think?

For a reference point, I've started the creation of Exports in the performance suite: https://github.com/nasa/openmct/pull/5236/files#diff-a991f8b4bf7b638ac578312c5b02de1b212962ffe834058d3c5da0a46c120009

**Reply by dimitriharding** [reply]
That sounds good actually.

**Reply by unlikelyzero** [reply]
@dimitriharding we did not receive your CLA. Can you tell us which address you sent it to?

**Reply by jvigliotta** [reply]
@dimitriharding I'll look into the status of your CLA, sometimes it can take a bit. I will keep you posted. Thanks for your patience!

**Reply by dimitriharding** [reply]
@jvigliotta / @unlikelyzero My CLA status should be good now, could we check on this?

**Reply by unlikelyzero** [reply]
@dimitriharding you are good to go!

**Reply by unlikelyzero** [reply]
@dimitriharding let me know how exportAsJson goes :)

**Comment by unlikelyzero** [comment]
@gholin you are the first! https://github.com/nasa/openmct/pull/5128

**Comment by MarkCodering** [comment]
@unlikelyzero It is Mark here. I sent my signed CLA based on guidelines from #3821 yesterday from the email <mark.chen.sstm@gmail.com>. Could you confirm the submission? Also, I am interested in this initiative, and could you check the CLA status so I can add my contribution to the codebase. Thanks for your help.

**Reply by jvigliotta** [reply]
@MarkCodering Hi Mark! It usually takes about a week turn around time for CLA's. Sometimes longer, but usually around a week at minimum. I'll keep an eye out and once we have it confirmed, we'll let you know!

**Reply by MarkCodering** [reply]
@jvigliotta Understood. But, I haven't received any updates from NASA yet. Do I need to submit any supplement material?

**Reply by MarkCodering** [reply]
Hi, @jvigliotta. A quick follow-up: Do we have any updates from NASA regarding CLA sign-up?

**Reply by jvigliotta** [reply]
@MarkCodering I am very sorry for dropping this, we do have your CLA on file.

**Comment by eddiejaoude** [comment]
Sounds interesting John! Debbie directed me here

**Reply by unlikelyzero** [reply]
@eddiejaoude that's great! Would you like to contribute?

**Reply by loftwah** [reply]
@eddiejaoude gave me the tip too :). I would love to contribute further.

**Reply by eddiejaoude** [reply]
Yes @unlikelyzero I would like to contribute and I am sure my community would too!

**Reply by loftwah** [reply]
I filled out the wrong CLA, it has been corrected now. [Here](https://nasa.github.io/openmct/static/files/ind-cla-open-mct.pdf) is the correct one for individual contributors.

**Comment by marcel-veselka** [comment]
Hi @unlikelyzero I am interested.

I read all about how to contribute and checked help wanted but didn't too much info on how and where I could start. 
Any advice on how to start? ... or at least start with learning before I will be ready to contribute?

I didn't submit CLA yet.

**Comment by TejAllam** [comment]
Hello John! I came across your talk on Playwright and wanted to contribute to the project. 
I have some experience with Playwright as part of my work. Would love to contribute to Open source projects. 
Please advice.

**Reply by unlikelyzero** [reply]
@TejAllam welcome aboard! If you haven't already, please submit the CLA. https://github.com/nasa/openmct/discussions/3821

We recommend that everyone run through the getting started documentation to get familiar with playwright, then Open MCT and playwright: https://github.com/nasa/openmct/tree/master/e2e#getting-started

Once you've done that, would you mind taking a look at https://github.com/nasa/openmct/issues/3213#issuecomment-1186831767 ? If you have an additional questions, we can follow up on that ticket!

**Reply by TejAllam** [reply]
Sure @unlikelyzero . Will do right away. Thanks!

**Comment by sabrinapowell** [comment]
Hi @unlikelyzero, I was recently introduced to the project by one of the contributors (Dimitri Harding). I'm really interested in contributing and would like some info on getting started. Please advice!

**Reply by unlikelyzero** [reply]
@sabrinapowell welcome aboard! If you haven't already, please submit the CLA. https://github.com/nasa/openmct/discussions/3821

We recommend that everyone run through the getting started documentation to get familiar with playwright, then Open MCT and playwright: https://github.com/nasa/openmct/tree/master/e2e#getting-started

Once you've done that, would you mind pinging me in this thread?

**Reply by sabrinapowell** [reply]
I've submitted the CLA, reviewed the docs and set up the project locally.

**Reply by unlikelyzero** [reply]
@sabrinapowell would you mind taking a stab at one of these https://github.com/nasa/openmct/blob/e4a36532e7b8cb36effdbb12ad2cd4398330aa0c/e2e/tests/functional/plugins/timeConductor/timeConductor.e2e.spec.js#L151-L169

**Reply by sabrinapowell** [reply]
Sure! I'll take the first one

**Reply by sabrinapowell** [reply]
@unlikelyzero I have opened an issue [here](https://github.com/nasa/openmct/issues/5905) and a [Draft PR here](https://github.com/nasa/openmct/pull/5906). Please take a look at the PR when you get a chance. I have noted a bug I am seeing on it.

**Reply by sabrinapowell** [reply]
@unlikelyzero I'm interested in picking up another area. Please let me know what's available

**Reply by ozyx** [reply]
Hi @sabrinapowell ! Thanks for your patience. A great place to pick up would be filling out some test stubs in our [notebook](https://github.com/nasa/openmct/blob/master/e2e/tests/functional/plugins/notebook/notebook.e2e.spec.js) or our [import](https://github.com/nasa/openmct/blob/master/e2e/tests/functional/plugins/importAndExportAsJSON/importAsJson.e2e.spec.js) and [export](https://github.com/nasa/openmct/blob/master/e2e/tests/functional/plugins/importAndExportAsJSON/exportAsJson.e2e.spec.js) e2e tests.

Let us know if you plan on picking any of these up and if we can provide any guidance to help you get started! Thanks again!

**Comment by saram7619** [comment]
Hi, I reviewed the docs, emailed the signed CLA two weeks ago and I have installed the project locally. I am wanting to contribute.

**Reply by eddiejaoude** [reply]
Awesome Sarah 💪

**Reply by ozyx** [reply]
@saram7619 It has been approved! Thanks for your patience. :)

**Reply by saram7619** [reply]
@ozyx Thanks for accepting my CLA, can’t wait to start contributing.

**Reply by ozyx** [reply]
Hi @saram7619 !

A great place to start would be with filling out some test stubs in our [notebook](https://github.com/nasa/openmct/blob/master/e2e/tests/functional/plugins/notebook/notebook.e2e.spec.js) or our [import](https://github.com/nasa/openmct/blob/master/e2e/tests/functional/plugins/importAndExportAsJSON/importAsJson.e2e.spec.js) and [export](https://github.com/nasa/openmct/blob/master/e2e/tests/functional/plugins/importAndExportAsJSON/exportAsJson.e2e.spec.js) e2e tests.

Let us know if you plan on picking any of these up and if we can provide any guidance to help you get started! Thanks!

**Reply by saram7619** [reply]
I’ll take a look tomorrow morning. Thanks for the opportunity to contribute.

**Reply by saram7619** [reply]
Hi @ozyx I'm planning to pick the test stubs in your notebook. Going to get started now.

**Comment by saram7619** [comment]
Hi @ozyx, I will take the ones in the notebook. Is it one test.fixme at a time?

**Reply by ozyx** [reply]
@saram7619 awesome! Please feel free to tackle as many or as few as you like in a single PR. Smaller PRs are generally quicker to review, however. Be sure to check out our [e2e testing README](https://github.com/nasa/openmct/blob/master/e2e/README.md) for some tips and feel free to reach out with any other questions you may have. Thanks for taking this on!
